service/src/java/org/apache/hive/service/cli/thrift/ThriftBinaryCLIService.java

```diff
@@ -113,43 +113,68 @@ protected void initServer() {
+
         @Override
         public ServerContext createContext(TProtocol input, TProtocol output) {
           Metrics metrics = MetricsFactory.getInstance();
           if (metrics != null) {
-            try {
-              metrics.incrementCounter(MetricsConstant.OPEN_CONNECTIONS);
-              metrics.incrementCounter(MetricsConstant.CUMULATIVE_CONNECTION_COUNT);
-            } catch (Exception e) {
-              LOG.warn("Error Reporting JDO operation to Metrics system", e);
-            }
+            metrics.incrementCounter(MetricsConstant.OPEN_CONNECTIONS);
+            metrics.incrementCounter(MetricsConstant.CUMULATIVE_CONNECTION_COUNT);
           }
           return new ThriftCLIServerContext();
         }
 
+        /**
+         * This is called by the Thrift server when the underlying client
+         * connection is cleaned up by the server because the connection has
+         * been closed.
+         */
         @Override
         public void deleteContext(ServerContext serverContext, TProtocol input, TProtocol output) {
           Metrics metrics = MetricsFactory.getInstance();
           if (metrics != null) {
-            try {
-              metrics.decrementCounter(MetricsConstant.OPEN_CONNECTIONS);
-            } catch (Exception e) {
-              LOG.warn("Error Reporting JDO operation to Metrics system", e);
-            }
+            metrics.decrementCounter(MetricsConstant.OPEN_CONNECTIONS);
           }
-          ThriftCLIServerContext context = (ThriftCLIServerContext) serverContext;
-          SessionHandle sessionHandle = context.getSessionHandle();
-          if (sessionHandle != null) {
-            LOG.info("Session disconnected without closing properly. ");
+
+          final ThriftCLIServerContext context = (ThriftCLIServerContext) serverContext;
+          final Optional<SessionHandle> sessionHandle = context.getSessionHandle();
+
+          if (sessionHandle.isPresent()) {
+            // Normally, the client should politely inform the server it is
+            // closing its session with Hive before closing its network
+            // connection. However, if the client connection dies for any reason
+            // (load-balancer round-robin configuration, firewall kills
+            // long-running sessions, bad client, failed client, timed-out
+            // client, etc.) then the server will close the connection without
+            // having properly cleaned up the Hive session (resources,
+            // configuration, logging etc.). That needs to be cleaned up now.
+            LOG.warn(
+                "Client connection bound to {} unexpectedly closed: closing this Hive session to release its resources. "
+                    + "The connection processed {} total messages during its lifetime of {}ms. Inspect the client connection "
+                    + "for time-out, firewall killing the connection, invalid load balancer configuration, etc.",
```

Comment:
A normal case like beeline exit with ctrl + c,  could we put this case here?

Reply:
Does beeline gracefully shutdown the connection on ctrl+c do you know?  It should :)

Reply:
I checked the codes, the session has already been gracefully shutdown in the shutdown hook of beeline,  we don't need to add this...